### PR TITLE
feat: Add loading indicator to search (fix #1219)

### LINKED ISSUE #1219: Implement a nice search result loading placeholder component
Instead of showing "Searching..." while we request and load more search results, we should show at least a spinner, but I think it'd be quite nice if we took a page out of the kind of wireframe-with-no-content loading screens on things like Facebook. The search result styles are consistent, so a nice animation with their boxes would make the loading screen possibly feel faster, if I'm making any sense.

### PR DESCRIPTION
**Pull request by tofumatt**
fixes #1219

### Video
feb-22-2017 00-51-11

### CODE REVIEW

**Comment by tofumatt** [comment]
This is ready for review but it's not a blocker, just had it ready and figured I'd push it.

Feel free to leave this until later if you're pressed for time (or someone else can r? it if they're bored)

File: src/ui/components/LoadingIndicator/LoadingIndicator.scss [hunk 1]
```diff
@@ -0,0 +1,17 @@
+.LoadingIndicator {
+  height: 48px;
+  width: 48px;
+
+  &.animating {
+    animation: spin 1.1s 1 cubic-bezier(0.42, 0.26, 0.04, 1.21);
```
Comment (kumar303) [inline src/ui/components/LoadingIndicator/LoadingIndicator.scss]: ooh fancy

File: src/ui/components/LoadingIndicator/index.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+  static defaultProps = {
+    children: null,
+    timeToReset: 500,
+  }
+
+  restartAnimation = (event) => {
+    const { timeToReset } = this.props;
+    const iconDOM = ReactDOM.findDOMNode(this.loadingIcon);
+    iconDOM.classList.remove('animating');
```
Comment (kumar303) [inline src/ui/components/LoadingIndicator/index.js]: Why is this needed? Can't you do it like this?
````js
restartAnimation = (event) => {
  ...
  this.setState({ animating: false });
}

render() {
  const className = classNames('LoadingIndicator', {
    animating: this.state.animating,
  });
  return <Icon className={className} name="fox" />;
}
````

File: src/amo/components/SearchResults.js [hunk 3]
```diff
@@ -29,10 +30,20 @@ class SearchResults extends React.Component {
...
     let messageText;
 
     if (loading) {
-      messageText = i18n.gettext('Searching...');
+      loadingMessage = (
+        <div className="LoadingIndicator-container"
+          ref={(ref) => { this.LoadingIndicatorContainer = ref; }}>
+          <LoadingIndicator>
+            <span className="visually-hidden">
+              {i18n.gettext('Searching…')}
```
Comment (kumar303) [inline src/amo/components/SearchResults.js]: is the ellipsis okay for localization?

File: src/amo/components/SearchResults.js [hunk 4]
```diff
@@ -29,10 +30,20 @@ class SearchResults extends React.Component {
...
+    let loadingMessage;
     let messageText;
 
     if (loading) {
-      messageText = i18n.gettext('Searching...');
+      loadingMessage = (
+        <div className="LoadingIndicator-container"
+          ref={(ref) => { this.LoadingIndicatorContainer = ref; }}>
+          <LoadingIndicator>
+            <span className="visually-hidden">
```
Comment (kumar303) [inline src/amo/components/SearchResults.js]: It seems like it might be common to include hidden text like this. I think this would be nicer as:
````js
<LoadingIndicator altText={i18n.gettext('Searching…')} />
````

File: src/ui/components/LoadingIndicator/index.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  }
+
+  static defaultProps = {
+    children: null,
+    timeToReset: 500,
+  }
+
+  restartAnimation = () => {
+    const { timeToReset } = this.props;
+    const iconDOM = ReactDOM.findDOMNode(this.loadingIcon);
```
Comment (kumar303) [inline src/ui/components/LoadingIndicator/index.js]: Why is this needed? Can't you do it like this?
````js
restartAnimation = (event) => {
  ...
  this.setState({ animating: false });
}

render() {
  const className = classNames('LoadingIndicator', {
    animating: this.state.animating,
  });
  return <Icon className={className} name="fox" />;
}
````

File: src/ui/components/LoadingIndicator/LoadingIndicator.scss (hunk 1 above)
Comment (tofumatt) [inline src/ui/components/LoadingIndicator/LoadingIndicator.scss]: I have @sarahmonster to thank 😄

File: src/amo/components/SearchResults.js (hunk 4 above)
Comment (tofumatt) [inline src/amo/components/SearchResults.js]: Oh that's nicer, agreed.

File: src/amo/components/SearchResults.js (hunk 3 above)
Comment (tofumatt) [inline src/amo/components/SearchResults.js]: Yeah, it's okay. The triple dots were there before and that's worse (and we use ellipsis elsewhere).

File: src/ui/components/LoadingIndicator/index.js (hunk 2 above)
Comment (tofumatt) [inline src/ui/components/LoadingIndicator/index.js]: Totally. Sometimes I forget about React state stuff but that's better AND easier to test. 👍

File: src/ui/components/LoadingIndicator/index.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    altText: null,
+    timeToReset: 500,
+  }
+
+  constructor(props) {
+    super(props);
+    this.state = { animating: true };
+  }
+
+  restartAnimation = () => {
```
Comment (mstriemer) [inline src/ui/components/LoadingIndicator/index.js]: You shouldn't need any JS to do this. The timing will be slightly different but roughly accurate to what you had.

```scss
.LoadingIndicator {
  animation: spin 1.6s infinite cubic-bezier(0.42, 0.26, 0.04, 1.21);
  height: 48px;
  width: 48px;
}

@keyframes spin {
  0% {
    transform: rotate(0deg);
  }

  66% {
    transform: rotate(720deg);
  }

  100% {
    transform: rotate(720deg);
  }
}

Comment (tofumatt) [inline src/ui/components/LoadingIndicator/index.js]: Oh shoot that's cool (and makes sense). Okay, I'll give that a go, cheers.

File: tests/client/ui/components/TestLoadingIndicator.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    assert.isTrue(root.state.animating);
+
+    Simulate.animationEnd(iconNode);
+
+    assert.isFalse(root.state.animating);
+
+    setTimeout(() => {
+      assert.isTrue(root.state.animating);
+      done();
+    }, 10);
```
Comment (kumar303) [inline tests/client/ui/components/TestLoadingIndicator.js]: These kind of tests are hard to write but doing it like this may fail occasionally since the timing is dependent on how fast the Travis VM is processing events. Also, slowing down our test suite by 10 milliseconds should be avoided when possible.

I think you have a few options:
1. Figure out how to return a promise from `restartAnimation()` and call it directly. You could add a separate test to make sure the `animationEnd` event is hooked up.
2. Use mock injection to replace `setTimeout` altogether. You could do it by passing it as a prop, for example. With this approach you'd have to trigger the callback manually, like `mockSetTimeout.firstCall.args[0]()` to simulate `setTimeout` usage.
3. Maybe if you move it all to CSS you can delete these tests entirely :)

Comment (tofumatt) [inline tests/client/ui/components/TestLoadingIndicator.js]: As @mstriemer mentioned seems it can be done in CSS; I'm just fixing up another PR then gonna do that here. Should tidy it up a million.

Comment (mstriemer) [inline tests/client/ui/components/TestLoadingIndicator.js]: These tests shouldn't be needed if this is done in CSS but sinon provides [fake timers](http://sinonjs.org/docs/#clock) that can be used to test code with `setTimeout()` synchronously if we need it elsewhere.

We're currently using them in [TestHoverIntent.js](https://github.com/mozilla/addons-frontend/blob/89ff059ead85a4c813d018388ed093d4a1b4eaf4/tests/client/core/components/TestHoverIntent.js#L51).

**Comment by tofumatt** [comment]
Ready for another r? This one is pure CSS, thanks @mstriemer

File: src/ui/components/LoadingIndicator/index.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+import './LoadingIndicator.scss';
+
+
+export default class LoadingIndicator extends React.Component {
+  static propTypes = {
+    altText: PropTypes.node,
+  }
+
+  static defaultProps = {
+    altText: null,
```
Comment (kumar303) [inline src/ui/components/LoadingIndicator/index.js]: I guess this could default to `gettext('Loading…')`, right?

File: tests/client/ui/components/TestLoadingIndicator.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+import LoadingIndicator from 'ui/components/LoadingIndicator';
+
+
+describe('<LoadingIndicator />', () => {
+  function render(props) {
+    return renderIntoDocument(<LoadingIndicator {...props} />);
+  }
+
+  it('renders alText', () => {
```
Comment (kumar303) [inline tests/client/ui/components/TestLoadingIndicator.js]: typo: altText

File: src/ui/components/LoadingIndicator/index.js (hunk 8 above)
Comment (tofumatt) [inline src/ui/components/LoadingIndicator/index.js]: True, actually, good call. 👍
